src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs

```diff
@@ -4408,5 +4408,65 @@ public async Task ResetAfterCompleteAsync_POSTWithResponseBodyAndTrailers_Reques
             Assert.Single(_decodedHeaders);
             Assert.Equal("Custom Value", _decodedHeaders["CustomName"]);
         }
+
+        [Fact]
+        public async Task HEADERS_Received_Latin1_AcceptedWhenLatin1OptionIsConfigured()
+        {
+            _serviceContext.ServerOptions.Latin1RequestHeaders = true;
+
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "GET"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+                // The HPackEncoder will encode £ as 0xA3 aka Latin1 encoding.
+                new KeyValuePair<string, string>("X-Test", "£"),
+            };
+
+            await InitializeConnectionAsync(context =>
+            {
+                Assert.Equal("£", context.Request.Headers["X-Test"]);
+                return Task.CompletedTask;
+            });
+
+            await StartStreamAsync(1, headers, endStream: true);
+
+            var headersFrame = await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 55,
+                withFlags: (byte)(Http2HeadersFrameFlags.END_HEADERS | Http2HeadersFrameFlags.END_STREAM),
+                withStreamId: 1);
+
+            await StopConnectionAsync(expectedLastStreamId: 1, ignoreNonGoAwayFrames: false);
+
+            _hpackDecoder.Decode(headersFrame.PayloadSequence, endHeaders: false, handler: this);
+
+            Assert.Equal(3, _decodedHeaders.Count);
+            Assert.Contains("date", _decodedHeaders.Keys, StringComparer.OrdinalIgnoreCase);
+            Assert.Equal("200", _decodedHeaders[HeaderNames.Status]);
+            Assert.Equal("0", _decodedHeaders["content-length"]);
+        }
+
+        [Fact]
+        public async Task HEADERS_Received_Latin1_RejectedWhenLatin1OptionIsNotConfigured()
+        {
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "GET"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+                // The HPackEncoder will encode £ as 0xA3 aka Latin1 encoding.
```

Comment:
It's interesting that the encoder didn't need to be modified.

Reply:
Yeah. I though I was going to have to use a custom HPackEncoder, but it turns out that it's just [casting chars to bytes](https://github.com/dotnet/aspnetcore/blob/master/src/Shared/Http2/Hpack/HPackEncoder.cs#L155). The response header's are normally [validated by Kestrel's HttpResponseHeaders dictionary](https://github.com/dotnet/aspnetcore/blob/4dd9bfc492452174f9329c3e1fc4807cb393c05a/src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs#L247-L267).